### DISCUSSION TITLE
Pi4J 2.6.0 initialisation not catchable

### DISCUSSION THREAD

**Discussion by Ir3nicuz** [Pi4J/pi4j#363] [Ideas]
Today I updated my program from pi4j 2.3.0 to 2.6.0 and i now i have the issue that i cannot catch content creation errors properly.
My IDE runs on windows, so I expect that before deploying to the Raspberry the pi4j content cannot be created.
I faced this by adding "throws Exception" to the initializer and it was good enough to test and debug the program without the hardware part before deploying and test the hardware part after deploying.

But with 2.6.0 the content creation or the output creation no longer throws a catchable exception. Instead i get a non-catchable "java.lang.ExceptionInInitializerError".
Is there another way to check the content on validity and relay that to an exception?

`
public final void initialize() throws Exception { 
        Pi4jContext = Pi4J.newAutoContext();
        DigitalOutput output = Pi4jContext.create(DigitalOutput.newConfigBuilder(Pi4jContext)
            .provider("pigpio-digital-output")
            .address(gpioNumber)
            .initial(initState)
            .shutdown(shutdownState));
}
`

**Comment by eitch** [comment]
Hi, that ExceptionInInitializerError is an error not an exception, and probably has something to do with native libraries. Can you show us the stack trace?

**Comment by taartspi** [comment]
I believe this user is attempting to run pi4j on their development machine, not on a Pi.  

Am I correct you are executing your code on a machine that is not an ARM based Pi ?

**Reply by Ir3nicuz** [reply]
> Am I correct you are executing your code on a machine that is not an ARM based Pi ?

Excactly. A exception is the expected reaction of the lib in this case on a x86 development machine.

2.3.0 does.
2.6.0 does not. 

Instead 2.6.0 bring up an error instead of an exception which cannot be catched by definition. 
If this is the intended behaviour of the lib, i ask for a property or function to check the validity of the context generation.
Or if not intended, a fix to be able to create the context target machine reliable again.

**Comment by taartspi** [comment]
I will ask @eitch or @FDelporte to comment if they see a different answer.  Pi4j-V2 targets only the Raspberry Pi family and no other architectures and IO structure.  So I do not believe we will create any stubs programs to support your test method.     Likely some change in java code, build process or native code resulted in the failure surfacing differently in 2.6

**Reply by Ir3nicuz** [reply]
I tried to setup a remote deploy and debug connection from the IDE to the raspberry without success
....first develoment concept failed

Developing direct on the raspberry is no suitable option if the hardware is not a standard desktop setup (mouse / keyboard / display ...)
....second development concept failed

Developing on an ordenary office machine is a valid concept in my oppinion. And 2.3.0 supports very fine and suitable with a exception.

But ok, if this is the intended behaviour i have to aquire a fourth development concept or roll back to 2.3.0

**Comment by eitch** [comment]
@Ir3nicuz can you supply the stack trace? At the moment i don't know what component is failing in your case, and can thus not offer a fix.

What i can say, is that we usually implement the pi4j in its own component using an interface, then one can simply detect that one is running on the desktop and use a different concrete class which does not use pi4j. Another concrete class can then be instantiated when running on the Raspberry Pi, which then communicates with the pi4j library.

**Reply by Ir3nicuz** [reply]
This is the stack trace if i try to run my program with pi4j 2.6.0 on my win10 development machine with Netbeans IDE:
[stacktrace.txt](https://github.com/user-attachments/files/15512749/stacktrace.txt)

**Reply by FDelporte** [reply]
@eitch could it be as simple as just adding `amd64` to the supported options? ;-) 

```
Pi4J has detected and UNKNOWN/UNSUPPORTED 'os.arch' : [amd64]; only 'arm|armhf' and 'arm64|aarch64' are supported
```

**Reply by eitch** [reply]
From the stack trace i don't think that will help, as then it will just lead to the next error, where it can't find the libraries, as there are no DLLs for pigpiod.

**Comment by Ir3nicuz** [comment]
I found this little bit hackish workaround but good enough for now (at least for windows where no "BoardModel" exists):

`String board = BoardInfoHelper.current().getBoardModel().getLabel();
        if (board == null || board.equals("Unknown")) {
            throw new IllegalStateException("not a pi");
        }`

Putting this before the context creation in my initialise method solved the problem.

**Comment by eitch** [comment]
In my opinion, it would be best, that you load the mock context when on a non-Raspberry Pi platform. See an example here: https://github.com/Pi4J/pi4j-template-javafx/blob/main/src/main/java/com/pi4j/mvc/util/Pi4JContext.java

**Reply by Ir3nicuz** [reply]
`runsOnPi()` does exactly what i'm looking for. -> A check of the system properties without throwing errors of missing dlls or native content.

I missed this because i just examined the context object creation and not the static Pi4JContext class possibilities. Thanks.

**Reply by FDelporte** [reply]
I applied this method in Spring Boot starter, but couldn't we integrate this in the library itself in `Pi4J.newAutoContext();` using the boardinfo?

Your proposed solution @eitch here: https://github.com/Pi4J/pi4j-springboot/blob/main/pi4j-spring-boot/src/main/java/com/pi4j/spring/boot/ContextConfiguration.java#L32

**Reply by eitch** [reply]
@FDelporte we certainly could.

**Comment by eitch** [comment]
I am glad you found a solution.
